### PR TITLE
Add krims and lazyten

### PR DESCRIPTION
**Pull request by mfherbst**
This PR adds the [krims](https://lazyten.org/krims) and [lazyten](https://lazyten.org) projects to spack.

Let me know if you see possible improvements.

### CODE REVIEW

File: var/spack/repos/builtin/packages/lazyten/package.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack import *
+import os
+
+
+class Lazyten(CMakePackage):
+    """FIXME: Put a proper description of your package here."""
```
Comment (davydden) [inline var/spack/repos/builtin/packages/lazyten/package.py]: Description

Comment (mfherbst) [reply var/spack/repos/builtin/packages/lazyten/package.py]: Good catch ... must have forgotten that one FIXME.

File: var/spack/repos/builtin/packages/krims/package.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    homepage = "http://lazyten.org/krims"
+    url      = "https://github.com/lazyten/krims/archive/v0.2.1.tar.gz"
+
+    version("0.2.1", "bf219656fd81c342a4782bad7d7beed0")
+    version("develop", git="https://github.com/lazyten/krims.git",
+            branch="master")
+
+    variant("build_type", default="DebugRelease",
+            description="The build type to build",
+            values=("Debug", "Release", "DebugRelease"))
```
Comment (junghans) [inline var/spack/repos/builtin/packages/krims/package.py]: This comes automatically from `CMakePackage`, so why did you add that? Does this package doesn't support `MinSizeRel`?

File: var/spack/repos/builtin/packages/krims/package.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    conflicts("%clang@:3.5")
+
+    def cmake_args(self):
+        spec = self.spec
+
+        args = [
+            # TODO Always disable tests for now
+            "-DKRIMS_ENABLE_TESTS=OFF",
+            # TODO Only build shared libs for now
+            "-DBUILD_SHARED_LIBS=ON",
```
Comment (junghans) [inline var/spack/repos/builtin/packages/krims/package.py]: Just make it a variant!

File: var/spack/repos/builtin/packages/krims/package.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            #
+            "-DAUTOCHECKOUT_MISSING_REPOS=OFF",
+            "-DDRB_MACHINE_SPECIFIC_OPTIM_Release=ON",  # Adds -march=native
+        ]
+
+        value = "ON" if "+examples" in spec else "OFF"
+        args.append("-DKRIMS_ENABLE_EXAMPLES=" + value)
```
Comment (junghans) [inline var/spack/repos/builtin/packages/krims/package.py]: `args.append("-DKRIMS_ENABLE_EXAMPLES=%s' % ('+examples' in spec)`

File: var/spack/repos/builtin/packages/lazyten/package.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    conflicts("%intel", msg="lazyten only builds with gcc and clang")
+    conflicts("%gcc@:4.8")
+    conflicts("%clang@:3.5")
+
+    depends_on("krims@develop", when="@develop")
+    depends_on("krims@0.2.1", when="@0.4.1")
+
+    depends_on("blas")
+    depends_on("lapack")
+    depends_on("armadillo@4.000:")
```
Comment (junghans) [inline var/spack/repos/builtin/packages/lazyten/package.py]: Why `000`? Isn't `4` enough?

File: var/spack/repos/builtin/packages/krims/package.py (hunk 2 above)
Comment (mfherbst) [reply var/spack/repos/builtin/packages/krims/package.py]: `krims` comes with the CMake module [DebugReleaseBuild](https://github.com/lazyten/krims#cmake-module-debugreleasebuild), which drops support for `RelWithDebInfo` as well as `MinSizeRel` in favour of a simultanious build of Debug and Release (`DebugRelease`). That system is also employed for `lazyten`, hence we override the `build_type` variant there as well.

File: var/spack/repos/builtin/packages/krims/package.py (hunk 3 above)
Comment (mfherbst) [reply var/spack/repos/builtin/packages/krims/package.py]: In principle I agree, but especially building the tests is not quite there yet as it requires some further dependencies (`[rapidcheck](https://github.com/emil-e/rapidcheck)`) which are currently not in Spack: I'll work on that some other day. With regards to `BUILD_SHARED_LIBS` you are probably right. I'll give that a try in a sec. Thanks.

File: var/spack/repos/builtin/packages/krims/package.py (hunk 4 above)
Comment (mfherbst) [reply var/spack/repos/builtin/packages/krims/package.py]: Hmm ... ok. I thought the `%` operator was kind of deprecated. What about
`args.append("-DKRIMS_ENABLE_EXAMPLES=" + str("+examples" in spec))`

File: var/spack/repos/builtin/packages/lazyten/package.py (hunk 5 above)
Comment (mfherbst) [reply var/spack/repos/builtin/packages/lazyten/package.py]: Armadillos documentation always uses version numbers like `n.xxx`, so I thought I'd adhere to that convention. I guess `4:` is easier to read.

Comment (davydden) [reply var/spack/repos/builtin/packages/lazyten/package.py]: in spack `4:` will be anything above `4`, including `4.1`. So that should work indeed

**Comment by davydden** [comment]
p.s. @mfherbst it's generally easier to do PRs off the stand-alone branch, i.e. `git checkout -b new_branch`. You don't need to do anything now, but after this PR is merged, your local `develop` and remote `develop` from spack will be incompatible. So  you would have to do hard-reset to be able to keep up with changes upstream.
